File: Set.js
class Set {
  constructor(args) {
    this.entries = [];
    this.size = 0;

    this.add = (item) => {
      let absent = true;
      for (let i = 0; i < this.size; i++) {
        if (this.entries[i] === item) absent = false
      }
      if (absent) {
        this.entries[this.size] = item
        this.size++
      }
    }

    this.remove = (item) => {
      let newSet = [];
      let newIndex = 0;
      let oldIndex = 0;
      while (oldIndex < this.size) {
        if (this.entries[oldIndex] !== item) {
          newSet[newIndex] = this.entries[oldIndex]
          newIndex++
          oldIndex++
        } else {
          oldIndex++
        }
      }
      this.entries = newSet;
      this.size--
    }
    
    this.contains = (item) => {
      let verdict = false
      for (let i = 0; i < this.size; i++) {
        if (this.entries[i] === item) {
          verdict = true
        }
      }
      return verdict
    }

    this.init = () => {
      let items = [...args]
      for (let item of items){
        this.add(item)
      }
    }

    this.init()
  }
}



let assert = require('assert');

//--------------------testing initial class object---------------------------
assert.deepStrictEqual(new Set('string').entries
  , ['s', 't', 'r', 'i', 'n', 'g'], 'init not rendering strings properly');

assert.deepStrictEqual(new Set(['one', 'two', 'three']).entries
  , ['one', 'two', 'three'], 'init not rendering arrays of strings properly');

assert.deepStrictEqual(new Set([true, false, NaN, undefined]).entries
  , [true, false, NaN, undefined], 'init not rendering arrays of booleans, NaN and undefined properly');

assert.deepStrictEqual(new Set([1, 2, 3]).entries
  , [1, 2, 3], 'init not rendering arrays of numbers properly');

assert.deepStrictEqual(new Set([[1], [2], [3]]).entries
  , [[1], [2], [3]], 'init not rendering arrays of numbers properly');

assert.deepStrictEqual(new Set([{one: 1}, {two: 2}, {three: 3}]).entries
  , [{one: 1}, {two: 2}, {three: 3}], 'init not rendering arrays of objects properly');

assert.deepStrictEqual(new Set([3, 2, 1]).entries
, [3, 2, 1], 'init not preserving insertion order');


//------------------testing set functionality-------------------------
var test = new Set(['one', 'two', 'three'])

test.add('four')
assert.deepStrictEqual(test.entries, ['one', 'two', 'three', 'four'], 'add method not functioning properly')

test.remove('four')
assert.deepStrictEqual(test.entries, ['one', 'two', 'three'], 'remove method not functioning properly')

assert.deepStrictEqual(test.contains('one'), true, 'contains method not functioning properly')
assert.deepStrictEqual(test.contains('five'), false, 'contains method not functioning properly')

File: set.py
class set:
    def __init__(self, args):
        self.entries = []
        self.size = 0
        for x in args:
            self.add(x)

    def add(self, item):
        absent = True
        for x in self.entries:
            if x == item:
                absent = False
        if absent:
            self.entries.append(item)
            self.size += 1

    def remove(self, item):
        new_set = []
        new_index = 0
        old_index = 0
        while old_index < self.size:
            if self.entries[old_index] != item:
                new_set.append(self.entries[old_index])
                new_index += 1
                old_index += 1
            else:
                old_index += 1
        self.entries = new_set
        self.size -= 1

    def contains(self, item):
        verdict = False
        for x in self.entries:
            if x == item:
                verdict = True
        return verdict

import unittest

    #-----------------testing set initialization---------------------------

class InitCases(unittest.TestCase):
    def test_string(self):
        self.assertEqual(set('string').entries, ['s', 't', 'r', 'i', 'n', 'g'])
    def test_list(self):
        self.assertEqual(set(['one', 'two', 'three']).entries, ['one', 'two', 'three'])
    def test_tuple(self):
        self.assertEqual(set(('one', 'two', 'three')).entries, ['one', 'two', 'three'])
    def test_list_special(self):
        self.assertEqual(set([True, False, None]).entries, [True, False, None])
    def test_list_int(self):
        self.assertEqual(set([1, 2, 3]).entries, [1, 2, 3])
    def test_list_list(self):
        self.assertEqual(set([[1], [2], [3]]).entries, [[1], [2], [3]])
    def test_list_dict(self):
        self.assertEqual(set([{'one': 1}, {'two': 2}, {'three': 3}]).entries, [{'one': 1}, {'two': 2}, {'three': 3}])

    #------------------testing set functionality-------------------------
class MethodCases(unittest.TestCase):
    def test_methods(self):
        test = set(['one', 'two', 'three'])

        test.add('four')
        self.assertEqual(test.entries, ['one', 'two', 'three', 'four'])

        test.remove('four')
        self.assertEqual(test.entries, ['one', 'two', 'three'])

        self.assertEqual(test.contains('one'), True)
        self.assertEqual(test.contains('five'), False)

if __name__ == '__main__':
    unittest.main()
